### PR TITLE
feat: enable strictNullChecks

### PR DESCRIPTION
**Pull request by andreialecu**
This PR enables strictNullChecks for the entire package.

Mostly tests were affected, but doing so revealed some issues, most of which I fixed.

The `isDocument` type guard is failing in strict mode, I'll add another commit that fixes it. Related to #249

### CODE REVIEW

File: src/internal/schema.ts [hunk 1]
```diff
@@ -65,7 +65,7 @@ export function _buildSchema<T, U extends AnyParamConstructor<T>>(
     sch = new Schema(schemas.get(name), schemaOptions);
   } else {
     sch = sch.clone();
-    sch.add(schemas.get(name));
+    sch.add(schemas.get(name)!);
```
Comment (hasezoey) [inline src/internal/schema.ts:68]: why is an extra `!` here after everything?

File: src/internal/schema.ts [hunk 2]
```diff
@@ -76,13 +76,13 @@ export function _buildSchema<T, U extends AnyParamConstructor<T>>(
       /** Get Metadata for PreHooks */
       const preHooks: IHooksArray[] = Reflect.getMetadata(DecoratorKeys.HooksPre, cl);
       if (Array.isArray(preHooks)) {
-        preHooks.forEach((obj) => sch.pre(obj.method, obj.func as EmptyVoidFn));
+        preHooks.forEach(obj => sch!.pre(obj.method, obj.func as EmptyVoidFn));
```
Comment (hasezoey) [inline src/internal/schema.ts:79]: why is an `!` needed here?

File: test/models/arrayValidators.ts [hunk 3]
```diff
@@ -17,36 +17,36 @@ export enum ArrayValidatorEnumNumber {
 export class ArrayValidators {
   // String-Array Values
   @arrayProp({ items: String, maxlength: 3 })
-  public maxLength?: string[];
+  public maxLength!: string[];
```
Comment (hasezoey) [inline test/models/arrayValidators.ts:20]: why are all the properties here changed to `!`?

File: test/models/user.ts [hunk 4]
```diff
@@ -56,7 +47,7 @@ export class User extends defaultClasses.FindOrCreate {
   public roles?: Role[];
 
   @prop()
-  public job?: Job;
+  public job!: Job;
```
Comment (hasezoey) [inline test/models/user.ts:50]: why are all properties here converted to `!` too?

File: test/tests/shouldAdd.test.ts [hunk 5]
```diff
@@ -28,13 +29,13 @@ it('should add a language and job using instance methods', async () => {
     languages: ['english'],
     uniqueId: 'unique-id'
   });
-  await user.addJob({ position: 'Dark Wizzard', title: 'Archmage' });
+  await user.addJob(new Job({ position: 'Dark Wizzard', title: 'Archmage' }));
```
Comment (hasezoey) [inline test/tests/shouldAdd.test.ts:32]: why did this get transformed into `new Job`?

**Comment by hasezoey** [comment]
note: i labeled it as "breaking" because it changed internal behavior quite a bit

**Comment by hasezoey** [comment]
> but doing so revealed some issues, most of which I fixed

could you point me to some of these issues?

**Comment by hasezoey** [comment]
i noticed you have an formatter that removes the spaces between curly-braces, i would like to add the tslint rule (`whitespace`), but i couldnt get it working in my vscode without something other interfering that removes them again (maybe somehow its `vscode.typescript-language-features`, but i dont know how to deactivate it for this rule ...)

**Comment by andreialecu** [comment]
> with all the formatting changes its hard to review
> 
> * thanks for `orFail`, i didnt know such an function existed as an query helper

It existed but it wasn't defined properly. I just got a PR merged to fix it earlier today: 
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44579
:)

File: src/internal/schema.ts (hunk 1 above)
Comment (andreialecu) [reply src/internal/schema.ts:68]: Because a call to a `map.get()` can return undefined if the key doesn't exist, but this line above:
https://github.com/typegoose/typegoose/pull/254/files/b0bb10c3b10b63e7fba3cad4b45248251984d74b#diff-47432c0fca2331f971956d7674048e26L61

Ensures that the key would be there, so `!` is an override to inform the compiler that the result cannot be undefined.

File: src/internal/schema.ts (hunk 2 above)
Comment (andreialecu) [reply src/internal/schema.ts:79]: Because `sch?` is originally defined as an optional parameter for the function, so it can be assumed to be undefined, but it is definitely assigned above (lines 64-69) so we inform the compiler of that.

File: test/models/arrayValidators.ts (hunk 3 above)
Comment (andreialecu) [reply test/models/arrayValidators.ts:20]: These are test models, I just made them required to not have extra checks for undefined in tests.

File: test/tests/shouldAdd.test.ts (hunk 5 above)
Comment (andreialecu) [reply test/tests/shouldAdd.test.ts:32]: I just refactored it to get rid of the workaround that was present in `job.ts` where `titleInUppercase` was defined as an optional function, which was just an overall bad practice. It only affects tests and makes things cleaner.

**Comment by andreialecu** [comment]
> note: i labeled it as "breaking" because it changed internal behavior quite a bit

There are no breaking changes, there are just minor refactorings but everything is the same.

**Comment by andreialecu** [comment]
> i noticed you have an formatter that removes the spaces between curly-braces, i would like to add the tslint rule (`whitespace`), but i couldnt get it working in my vscode without something other interfering that removes them again (maybe somehow its `vscode.typescript-language-features`, but i dont know how to deactivate it for this rule ...)

I would strongly suggest to set up the prioject configuration in such a way that there is a lint that ensures the code style you want is preserved. 

The current formatting is a direct effect of settings in the project, I just have vscode autoformat it because `.vscode/settings.json` actually has `"editor.formatOnSave": true,` in it. It's basically the project's problem, they're not even my custom formatting rules. :)

**Comment by andreialecu** [comment]
> > but doing so revealed some issues, most of which I fixed
> 
> could you point me to some of these issues?

Basically it was possible some variables could end up being undefined.

Some additional checks here: 

https://github.com/typegoose/typegoose/pull/254/files#diff-17dec77f694b959d451c11a845c30a77R381-R463

And I verified all the places where I had to add `!` that the variables are definitely assigned in the flow above, for the core parts of the library.

For the tests, adding `!` doesn't change anything. The test code will fail regardless if something is wrong at runtime.

Basically all these changes are for `compile` time purposes, the built library is not any different.

**Comment by hasezoey** [comment]
> I would strongly suggest to set up the project configuration in such a way that there is a lint that ensures the code style you want is preserved.
> 
> The current formatting is a direct effect of settings in the project, I just have vscode autoformat it because .vscode/settings.json actually has "editor.formatOnSave": true, in it. It's basically the project's problem, they're not even my custom formatting rules. :)

sorry, this wasnt meant to be directed at you, i wanted to add some rules (like `whitespace`) but something i cant figure out is interfering with it and changes it
example of what i mean:
```ts
// having "whitespace" (for everything) on
// code to format:
class Something {}

// tslint --fix / "Fix all TSLint Problems" result in what is expected
class Something { }

// but running "Ctrl+S" (and having format on save enabled) or running "Format Document" results in
// shortly the expected format, but quickly changes back before actually saving
// result:
class Something {}

// and i just use the default formatter, where i just dont understand why it ignores the configs
```

**Comment by andreialecu** [comment]
I fixed `Ref` again, now `isDocument` works in strict mode. It was broken completely before.

File: src/types.ts [hunk 6]
```diff
@@ -268,7 +268,7 @@ export type RefSchemaType =
  * Reference another Model
  */
 // export type Ref<R, T extends RefType = mongoose.Types.ObjectId> = R | T; // old type, kept for easy revert
-export type Ref<R, T extends RefType = R extends { _id?: RefType } ? NonNullable<R['_id']> : mongoose.Types.ObjectId> = R | T;
+export type Ref<R, T extends RefType = (R extends { _id?: RefType } ? NonNullable<R['_id']> : mongoose.Types.ObjectId) | undefined> = R | T;
```
Comment (hasezoey) [inline src/types.ts:271]: what does this extra `| undefined` do here?

Comment (andreialecu) [reply src/types.ts:271]: Some more context here:
https://github.com/microsoft/TypeScript/issues/38323#issuecomment-626183067

Assume this code:

```ts
declare const ref: Ref<SomeOtherModel> | undefined;

if (isDocument(ref)) {
```

Original `Ref` type:
```ts
export type Ref<R, T extends RefType = R extends { _id?: RefType } ? NonNullable<R['_id']> : mongoose.Types.ObjectId> = R | T;
```

If `ref` is `undefined`, then `Ref<R, T>` works with two `undefined` values for both `R` and `T`, but `R` would not work with `undefined` because `R extends { _id?: RefType } ? NonNullable<R['_id']> : mongoose.Types.ObjectId` cannot ever be `undefined` by that definition.

So it just adds it as a possible value for `R`, which is valid for the case I mentioned above.

File: src/internal/schema.ts (hunk 1 above)
Comment (hasezoey) [reply src/internal/schema.ts:68]: thanks

**Comment by andreialecu** [comment]
> > I would strongly suggest to set up the project configuration in such a way that there is a lint that ensures the code style you want is preserved.
> > The current formatting is a direct effect of settings in the project, I just have vscode autoformat it because .vscode/settings.json actually has "editor.formatOnSave": true, in it. It's basically the project's problem, they're not even my custom formatting rules. :)
> 
> sorry, this wasnt meant to be directed at you, i wanted to add some rules (like `whitespace`) but something i cant figure out is interfering with it and changes it
> example of what i mean:
> 
> ```ts
> // having "whitespace" (for everything) on
> // code to format:
> class Something {}
> 
> // tslint --fix / "Fix all TSLint Problems" result in what is expected
> class Something { }
> 
> // but running "Ctrl+S" (and having format on save enabled) or running "Format Document" results in
> // shortly the expected format, but quickly changes back before actually saving
> // result:
> class Something {}
> 
> // and i just use the default formatter, where i just dont understand why it ignores the configs
> ```

Sure, I understand. Feel free to take over and re-run lint. 

Running `npm run lint -- --fix` does absolutely nothing to it for me. :)

Maybe a [husky](https://www.npmjs.com/package/husky) hook can be added to check lint before pushing.

**Comment by hasezoey** [comment]
>  Feel free to take over and re-run lint.

what i meant with that message is that i *would like* to add these rules, but because vscode's typescript formatter dosnt apply (resets) these rules i cannot add them

---

`npm run lint -- --fix` does nothing because the latest master isnt merged (which added some lint rules)

**Comment by andreialecu** [comment]
Sure, I'll be away for a bit though. So if you want to get it moving, feel free to take over and reformat it. :)

Again. There should be no breaking changes introduced by this. It just does **very minor** reorganizaton and strengthens the code, especially good as a base for future contributions.

**Comment by hasezoey** [comment]
> Again. There should be no breaking changes introduced by this. It just does very minor reorganizaton and strengthens the code, especially good as a base for future contributions.

i added this because there were so many changes and i didnt look at them at detail (because they were mostly formatting related changes) and because i had some questions
-> i will reevaluate the label once i looked at it in more detail, but it will not change anything that it will be in an minor (semver) update and not in an patch (semver) update

**Comment by hasezoey** [comment]
@andreialecu im currently checking out your pr, and i noticed that after switching to this pr, my vscode wants to *add* an space between `{}` again, i have no clue why it wants to do it now

**Comment by hasezoey** [comment]
@andreialecu do you think this pr is ready to merge?

**Comment by andreialecu** [comment]
Yes. I have no pending work on it. Go ahead. :)

**Comment by hasezoey** [comment]
> Yes. I have no pending work on it. Go ahead. :)

just asked because of the changes

---

if you have nothing more to add & the changes are fine, then i would merge it after the checks pass
